client/asset/btc/electrum_client.go

```diff
@@ -0,0 +1,1289 @@
+		if err != nil {
+			return nil, err
+		}
+		privKey, compressed, err := ew.privKeyForAddress(addr)
+		if err != nil {
+			return nil, err
+		}
+		// NewAddressWitnessPubKeyHash
+
+		// witnessProgram, err := txscript.PayToAddrScript(p2wkhAddr)
+
+		// hashCache := txscript.NewTxSigHashes(tx)
+
+		// txIn.Witness, err = txscript.WitnessSignature
+	}
+	*/
+
+	packet, err := psbt.NewFromUnsignedTx(inTx)
+	if err != nil {
+		return nil, err
+	}
+	psbtB64, err := packet.B64Encode()
+	if err != nil {
+		return nil, err
+	}
+
+	signedB, err := ew.wallet.SignTx(ew.walletPass(), psbtB64)
+	if err != nil {
+		return nil, err
+	}
+	return msgTxFromBytes(signedB)
+}
+
+type hash160er interface {
+	Hash160() *[20]byte
+}
+
+type pubKeyer interface {
+	PubKey() *btcec.PublicKey
+}
+
+func (ew *electrumWallet) privKeyForAddress(addr string) (*btcec.PrivateKey, error) {
+	addrDec, err := ew.decodeAddr(addr, ew.chainParams)
+	if err != nil {
+		return nil, err
+	}
+	wifStr, err := ew.wallet.GetPrivateKeys(ew.walletPass(), addr)
+	if err != nil {
+		return nil, err
+	}
+	wif, err := btcutil.DecodeWIF(wifStr)
+	if err != nil {
+		return nil, err
+	} // wif.PrivKey is the result
+
+	// Sanity check that PrivKey corresponds to the pubkey(hash).
+	var pkh []byte
+	switch addrT := addrDec.(type) { // maybe switch on interfaces like Hash160() and PubKey() instead of concrete types
+	case pubKeyer: // *btcutil.AddressPubKey:
+		// Get same format as wif.SerializePubKey()
+		var pk []byte
+		if wif.CompressPubKey {
+			pk = addrT.PubKey().SerializeCompressed()
+		} else {
+			pk = addrT.PubKey().SerializeUncompressed()
+		}
+		pkh = btcutil.Hash160(pk) // addrT.ScriptAddress() would require SetFormat(compress/uncompress)
+	case *btcutil.AddressScriptHash, *btcutil.AddressWitnessScriptHash:
+		return wif.PrivKey, nil // assume unknown redeem script references this pubkey
+	case hash160er: // p2pkh and p2wpkh
+		pkh = addrT.Hash160()[:]
+	}
+	wifPKH := btcutil.Hash160(wif.SerializePubKey())
+	if !bytes.Equal(pkh, wifPKH) {
+		return nil, errors.New("pubkey mismatch")
+	}
+	return wif.PrivKey, nil
+}
+
+func (ew *electrumWallet) walletPass() string {
+	ew.passMtx.RLock()
+	defer ew.passMtx.RUnlock()
+	return ew.pass
+}
+
+func (ew *electrumWallet) walletUnlock(pw []byte) error {
+	addr, err := ew.wallet.GetUnusedAddress()
+	if err != nil {
+		return err
+	}
+	pass := string(pw)
+	wifStr, err := ew.wallet.GetPrivateKeys(pass, addr)
+	if err != nil {
+		return err
+	} // that should be enough, but validate the returned keys in case they are empty or something
+	if _, err = btcutil.DecodeWIF(wifStr); err != nil {
+		return err
+	}
+	ew.passMtx.Lock()
+	ew.pass = pass
+	ew.passMtx.Unlock()
+	return nil
+}
+
+func (ew *electrumWallet) walletLock() error {
+	ew.passMtx.Lock()
+	ew.pass = ""
+	ew.passMtx.Unlock()
+	return nil
+}
+
+func (ew *electrumWallet) locked() bool {
+	ew.passMtx.RLock()
+	defer ew.passMtx.RUnlock()
+	return ew.pass == ""
```

Comment:
I was able to create the electrum wallet without a password.

Reply:
Oh!  I guess I'll change this scheme